tests/server_send.cpp

```diff
@@ -0,0 +1,349 @@
+			// following error:
+			// error: cannot bind packed field ... to int&
+			BH_LOG(log, DNET_LOG_DEBUG,
+					"ssend_test: "
+					"key: %s, backend: %d, user_flags: %llx, ts: %lld.%09lld, status: %d, size: %lld, "
+					"iterated_keys: %lld/%lld",
+				dnet_dump_id_len_raw(it->reply()->key.id, DNET_ID_SIZE, buffer),
+				(int)it->command()->backend_id,
+				(unsigned long long)it->reply()->user_flags,
+				(unsigned long long)it->reply()->timestamp.tsec, (unsigned long long)it->reply()->timestamp.tnsec,
+				(int)it->reply()->status, (unsigned long long)it->reply()->size,
+				(unsigned long long)it->reply()->iterated_keys, (unsigned long long)it->reply()->total_keys);
+#endif
+
+			copied++;
+		}
+
+		BH_LOG(log, DNET_LOG_NOTICE, "ssend_test: %s: dst_groups: %s, copied: %d",
+				id.to_string(),
+				print_groups(dst_groups), copied);
+
+		return copied;
+	};
+
+	int copied = 0;
+	std::vector<dnet_raw_id> ids = ssend_ids(s);
+	for (const auto &id: ids) {
+		copied += run_over_single_backend(s, id, dst_groups, iflags);
+	}
+
+	BOOST_REQUIRE_EQUAL(copied, num);
+}
+
+static void ssend_test_server_send(session &s, int num, const std::string &id_prefix, const std::string &data_prefix,
+		const std::vector<int> &dst_groups, uint64_t iflags)
+{
+	logger &log = s.get_logger();
+
+	s.set_trace_id(rand());
+	std::vector<std::string> keys;
+	for (int i = 0; i < num; ++i) {
+		std::string id = id_prefix + lexical_cast(i);
+		std::string data = data_prefix + lexical_cast(i);
+
+		ELLIPTICS_REQUIRE(res, s.write_data(id, data, 0));
+
+		keys.push_back(id);
+	}
+
+	BH_LOG(log, DNET_LOG_NOTICE, "%s: keys: %d, dst_groups: %s, starting copy",
+			__func__, num, print_groups(dst_groups));
+
+	//char buffer[2*DNET_ID_SIZE + 1] = {0};
+
+	int copied = 0;
+	auto iter = s.server_send(keys, iflags, dst_groups);
+	for (auto it = iter.begin(), iter_end = iter.end(); it != iter_end; ++it) {
+#if 0
+		// we have to explicitly convert all members from dnet_iterator_response
+		// since it is packed and there will be alignment issues and
+		// following error:
+		// error: cannot bind packed field ... to int&
+		BH_LOG(log, DNET_LOG_DEBUG,
+				"ssend_test: "
+				"key: %s, backend: %d, user_flags: %llx, ts: %lld.%09lld, status: %d, size: %lld, "
+				"iterated_keys: %lld/%lld",
+			dnet_dump_id_len_raw(it->reply()->key.id, DNET_ID_SIZE, buffer),
+			(int)it->command()->backend_id,
+			(unsigned long long)it->reply()->user_flags,
+			(unsigned long long)it->reply()->timestamp.tsec, (unsigned long long)it->reply()->timestamp.tnsec,
+			(int)it->reply()->status, (unsigned long long)it->reply()->size,
+			(unsigned long long)it->reply()->iterated_keys, (unsigned long long)it->reply()->total_keys);
+#endif
+
+		copied++;
+	}
+
+	BH_LOG(log, DNET_LOG_NOTICE, "%s: keys: %d, dst_groups: %s, copied total: %d",
+			__func__, num, print_groups(dst_groups), copied);
+
+	BOOST_REQUIRE_EQUAL(copied, num);
+}
+
+static bool ssend_register_tests(test_suite *suite, node &n)
```

Comment:
>You do not really help to make this code better with this. You stop progress.
I wish we spent this time searching for bugs.

Really? We've already pointed you on potential bugs and that your tests don't cover cases which lead to data lost. Review is the time to stop, understand and analyze code of new feature. And you don't help us with your ironic and reluctance to split this 2k+-lines patch where mixed new feature and unrelated refactoring and bugs fixes.

> I'm curious how many tests you will write for the next PR.
Let's not measure the number of tests, but the quality. And again, please, avoid using such intonation in your words, it doesn't make anything.

>where it test or guarantee that plain iterator doesn't remove data? Why didn't you implement this test for iterator?
Plain iterator doesn't remove data in any case with any flags and any types, there is no code which could remove data in plain iterator. Copy iterator with certain flag removes data, so it contains code which removes data and it should be tested that copy iterator doesn't remove data without this certain flag.